cmake/CeleritasUtils.cmake

```diff
@@ -43,15 +43,180 @@ macro(celeritas_find_package_config _package)
+    INTERFACE_INCLUDE_DIRECTORIES "${target_interface_include_directories}")
+
+  get_target_property(target_LINK_LIBRARIES ${target}
+    LINK_LIBRARIES
+  )
+  get_target_property(target_INTERFACE_LINK_LIBRARIES ${target}
+    INTERFACE_LINK_LIBRARIES )
+  set_target_properties(${target}_static PROPERTIES
+    LINK_LIBRARIES "${target_LINK_LIBRARIES}"
+    INTERFACE_LINK_LIBRARIES "${target_INTERFACE_LINK_LIBRARIES}"
+  )
+
+  set_target_properties(${target}_static PROPERTIES
+    POSITION_INDEPENDENT_CODE ON
+  )
+endif()
+
 endfunction()
 
+
+define_property(TARGET PROPERTY CELERITAS_CUDA_LIBRARY_TYPE
+  BRIEF_DOCS "Indicate the type of cuda library (STATIC and SHARED for nvlink usage, FINAL for linking into not cuda library/executable"
+  FULL_DOCS "Indicate the type of cuda library (STATIC and SHARED for nvlink usage, FINAL for linking into not cuda library/executable"
+)
+define_property(TARGET PROPERTY CELERITAS_CUDA_FINAL_LIBRARY
+  BRIEF_DOCS "Name of the final library corresponding to this cuda library"
+  FULL_DOCS "Name of the final library corresponding to this cuda library"
+)
+define_property(TARGET PROPERTY CELERITAS_CUDA_STATIC_LIBRARY
+  BRIEF_DOCS "Name of the static library corresponding to this cuda library"
+  FULL_DOCS "Name of the static library corresponding to this cuda library"
+)
+define_property(TARGET PROPERTY CELERITAS_CUDA_MIDDLE_LIBRARY
+  BRIEF_DOCS "Name of the shared (without nvlink step) library corresponding to this cuda library"
+  FULL_DOCS "Name of the shared (without nvlink step) library corresponding to this cuda library"
+)
+
+function(celeritas_add_library target)
+  if(NOT BUILD_SHARED_LIBS OR NOT CELERITAS_USE_CUDA)
+    list(SUBLIST ARGV 1 -1 NEWARGV)
+    add_library(${target} ${NEWARGV})
+    add_library(${target}_final ALIAS ${target})
+    add_library(${target}_cuda ALIAS ${target})
+    add_library(${target}_static ALIAS ${target})
+    return()
+  endif()
+
+  list(SUBLIST ARGV 1 -1 NEWARGV)
+
+  add_library(${target}_objects OBJECT ${NEWARGV})
+  add_library(${target}_static STATIC $<TARGET_OBJECTS:${target}_objects>)
+  add_library(${target}_cuda SHARED $<TARGET_OBJECTS:${target}_objects>)
+  # We need to use a dummy file as a library (per cmake) needs to contains
+  # at least one source file.  The real content of the library will be
+  # the cmake_device_link.o resulting from the execution of `nvcc -dlink`
+  # Also non-cuda related test, for example `gtest_detail_Macros`,
+  # will need to be linked again libceleritas_final while a library
+  # that the detends on and that uses Celeritas::Core (for example
+  # libCeleritasTest.so) will need to be linked against `libceleritas_cuda`.
+  # If both the `_cuda` and `_final` contains the `.o` files we would
+  # then have duplicated symbols (Here the symptoms will a crash
+  # during the cuda library initialization rather than a link error).
+  add_library(${target}_final SHARED ../src/base/dummy.cu)
```

Comment:
Since this is only needed for "cmake" and not for source, and since I think this relative path may cause problems, it would be best to (perhaps) generate the dummy file dynamically inside the build directory.  An easier but also acceptable solution would be to put `dummy.cu` inside `celeritas/cmake/`, and refer to it here as `${PROJECT_SOURCE_DIR}/cmake/dummy.cu` to avoid the relative paths.

Reply:
> it would be best to (perhaps) generate the dummy file dynamically inside the build directory.

Do you have sample (modern cmake) code for that?

Reply:
Since `dummy.cu` is essentially part of the `CeleritasUtils` implementation, why not have it alongside `CeleritasUtils.cmake` and then implement the pattern for CMake < 3.17:

```cmake
# At the top of CeleritasUtils.cmake
set(_CELERITAS_UTILS_DIR "${CMAKE_CURRENT_LIST_DIR}")

...

add_library(${target}_final SHARED "${_CELERITAS_UTILS_DIR}/dummy.cu")
```

or in CMake 3.17 and newer:

```cmake
add_library(${target}_final SHARED "${CMAKE_CURRENT_FUNCTION_LIST_DIR}/dummy.cu")
```

The [`CMAKE_CURRENT_FUNCTION_LIST_DIR`](https://cmake.org/cmake/help/latest/variable/CMAKE_CURRENT_FUNCTION_LIST_DIR.html) is recommended, but the first method is reliable enough until 3.17 is the minimum required.

Reply:
My suggested alternative since the CUDA stub is just a simple empty file, would be to create one per target in the `blah/CMakeFiles/` directory as needed:
```cmake
set(_stub "${CMAKE_CURRENT_BINARY_DIR}${CMAKE_FILES_DIRECTORY}/${_target}.cu")
file(WRITE "${_stub}" "/* stub */")
add_library(${target}_final "${_stub}"
```